keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,903 @@
+pool in the API server and thus make it available to the scheduler. In
+the future it might get extended to also store other information like
+health of the pool.
+
+### Identifying storage pools
+
+A method for identifying storage pools accessible via a CSI driver
+based on the existing topology information gets outlined below. The
+`external-provisioner` will be extended to handle the management of
+`CSIStoragePool` objects.
+
+For local storage in the node, there will be one such `CSIStoragePool`
+object per node and CSI driver on that node. For network attached
+storage, there will be one `CSIStoragePool` per CSI
+[Topology](https://github.com/container-storage-interface/spec/blob/4731db0e0bc53238b93850f43ab05d9355df0fd9/lib/go/csi/csi.pb.go#L1662-L1691)
+and driver.
+
+### Size of ephemeral inline volumes
+
+Currently Kubernetes has no information about the size of an ephemeral
+inline volume. A new `CSIVolumeSource.fsSize` field needs to be added
+to expose that in a vendor-agnostic way. Details for that are in
+https://github.com/kubernetes/enhancements/pull/1353.
+
+### Pod scheduling
+
+The Kubernetes scheduler monitors `CSIStoragePool` objects and
+excludes nodes with insufficient remaining capacity when it comes to
+making scheduling decisions for a pod which uses ephemeral inline
+volumes or persistent volumes with delayed binding. If the driver does
+not indicate that it supports capacity reporting, then the scheduler
+proceeds just as it does now, so nothing changes for existing CSI
+driver deployments.
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
+
+#### CSIStoragePool
+
+The prefix “CSI” is used to avoid confusion with other types that have
+a similar name and because some aspects are specific to CSI.
+
+```
+// CSIStoragePool identifies one particular storage pool and
+// stores the corresponding attributes. The spec is read-only.
+//
+// There should never be two objects with the same spec. If
+// a consumer of this information finds two objects with the
+// same spec, it should use the information from the one with the
+// most recent CreationTimestamp.
+type CSIStoragePool struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning and just has to
+	// meet the usual requirements (length, characters, unique). To ensure that
+    // there are no conflicts with other CSI drivers on the cluster, the recommendation
+    // is to use sp-<uuid>.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	Spec   CSIStoragePoolSpec
+	Status CSIStoragePoolStatus
+}
+
+// CSIStoragePoolSpec contains the constant attributes of a CSIStoragePool.
+//
+// A pool might only be accessible from a subset of the nodes in the
+// cluster. That subset can be identified either via NodeTopology or
+// Nodes, but not both. If neither is set, the pool is assumed
```

Comment:
If that's the case, you can have a NodeSelection struct which contains both NodeTopology and Nodes, and add validation to make sure only one is set.

Reply:
That validation can (and should) also be done for the `CSIStoragePool`. It seems simpler to me to avoid introducing a `NodeSelection` struct, but I don't have a strong opinion about this.